File: jvm/examples/io/github/amaye15/jsontoolsrs/examples/FeatureCombinations.java
package io.github.amaye15.jsontoolsrs.examples;

import io.github.amaye15.jsontoolsrs.JsonTools;
import io.github.amaye15.jsontoolsrs.JsonToolsHandle;

/**
 * Curated multi-feature pipelines, building on {@link FeatureByFeature}. Not an
 * exhaustive combinatorial sweep (the builder has ~10 independent toggles, so a literal
 * power-set would be 1000+ cases) -- these are realistic groupings of features commonly
 * used together, plus one "kitchen sink" example exercising nearly everything at once.
 * Mirrors ../../../../../../../../examples/feature_combinations.rs and
 * python/examples/feature_combinations.py.
 *
 * <p>Run with:
 *
 * <pre>
 * mvn -P examples compile exec:java \
 *     -Dexec.mainClass=io.github.amaye15.jsontoolsrs.examples.FeatureCombinations
 * </pre>
 */
public final class FeatureCombinations {

    private FeatureCombinations() {}

    public static void main(String[] args) {
        System.out.println("JSON Tools RS - Feature Combinations");
        System.out.println("=====================================\n");

        // 1. separator + lowercaseKeys + keyReplacement + handleKeyCollision
        System.out.println("1. separator + lowercaseKeys + keyReplacement + handleKeyCollision");
        String input = "{\"User\":{\"Full_Name\":\"John\"},\"Admin\":{\"Full_Name\":\"Jane\"}}";
        try (JsonToolsHandle tools = JsonTools.builder()
                .flatten()
                .separator("_")
                .lowercaseKeys(true)
                .keyReplacement("r'^(user|admin)_'", "")
                .handleKeyCollision(true)
                .build()) {
            System.out.println("   In:  " + input);
            System.out.println("   Out: " + tools.execute(input) + "\n");
        }

        // 2. keyReplacement + valueReplacement together
        System.out.println("2. keyReplacement + valueReplacement");
        input = "{\"usr_nm\":\"John\",\"usr_eml\":\"john@old.com\"}";
        try (JsonToolsHandle tools = JsonTools.builder()
                .flatten()
                .keyReplacement("usr_", "user_")
                .valueReplacement("@old.com", "@new.com")
                .build()) {
            System.out.println("   In:  " + input);
            System.out.println("   Out: " + tools.execute(input) + "\n");
        }

        // 3. All four empty-value filters together
        System.out.println(
                "3. removeEmptyStrings + removeNulls + removeEmptyObjects + removeEmptyArrays");
        input = "{\"name\":\"John\",\"bio\":\"\",\"age\":null,\"tags\":[],\"meta\":{}}";
        try (JsonToolsHandle tools = JsonTools.builder()
                .flatten()
                .removeEmptyStrings(true)
                .removeNulls(true)
                .removeEmptyObjects(true)
                .removeEmptyArrays(true)
                .build()) {
            System.out.println("   In:  " + input);
            System.out.println("   Out: " + tools.execute(input) + "\n");
        }

        // 4. Real-world normalization pipeline: replacement + filtering + collision
        System.out.println("4. lowercaseKeys + keyReplacement + filtering + handleKeyCollision");
        input = "{\"User_Name\":\"John\",\"User_Bio\":\"\",\"Admin_Name\":\"Jane\",\"Admin_Bio\":null}";
        try (JsonToolsHandle tools = JsonTools.builder()
                .flatten()
                .lowercaseKeys(true)
                .keyReplacement("r'^(user|admin)_'", "")
                .removeEmptyStrings(true)
                .removeNulls(true)
                .handleKeyCollision(true)
                .build()) {
            System.out.println("   In:  " + input);
            System.out.println("   Out: " + tools.execute(input) + "\n");
        }

        // 5. unflatten + keyReplacement + valueReplacement
        System.out.println("5. unflatten + separator + keyReplacement + valueReplacement");
        input = "{\"PREFIX_user_name\":\"john@OLD.com\",\"PREFIX_user_age\":30}";
        try (JsonToolsHandle tools = JsonTools.builder()
                .unflatten()
                .separator("_")
                .keyReplacement("PREFIX_user_", "profile_")
                .valueReplacement("@OLD.com", "@new.com")
                .build()) {
            System.out.println("   In:  " + input);
            System.out.println("   Out: " + tools.execute(input) + "\n");
        }

        // 6. normal mode + autoConvertTypes + valueReplacement + filtering
        System.out.println("6. normal + autoConvertTypes + valueReplacement + removeEmptyStrings");
        input = "{\"user\":{\"status\":\"ACTIVE\",\"note\":\"\",\"score\":\"95.5\"}}";
        try (JsonToolsHandle tools = JsonTools.builder()
                .normal()
                .autoConvertTypes(true)
                .valueReplacement("r'^ACTIVE$'", "enabled")
                .removeEmptyStrings(true)
                .build()) {
            System.out.println("   In:  " + input);
            System.out.println("   Out: " + tools.execute(input));
            System.out.println(
                    "   Note: nesting preserved, string replaced, empty note dropped, score converted\n");
        }

        // 7. Batch processing + parallel tuning + type conversion together
        System.out.println(
                "7. batch execute + parallelThreshold + numThreads + "
                        + "nestedParallelThreshold + autoConvertTypes");
        String[] batch = new String[150];
        for (int i = 0; i < batch.length; i++) {
            batch[i] = "{\"id\":\"" + i + "\",\"active\":\"true\"}";
        }
        try (JsonToolsHandle tools = JsonTools.builder()
                .flatten()
                .parallelThreshold(50)
                .numThreads(4)
                .nestedParallelThreshold(100)
                .autoConvertTypes(true)
                .build()) {
            String[] results = tools.executeBatch(batch);
            System.out.println("   Processed " + results.length + " documents");
            System.out.println("   Sample: " + results[0] + "\n");
        }

        // 8. Kitchen sink: (almost) every feature at once, on a realistic messy batch
        System.out.println("8. Kitchen sink - every applicable feature combined");
        String[] apiBatch = {
            "{\"API_Response\":{\"User_Data\":{\"First_Name\":\"John\",\"Email\":\"john@old.com\","
                    + "\"Bio\":\"\",\"Score\":\"88.5\"}}}",
            "{\"API_Response\":{\"User_Data\":{\"First_Name\":\"Jane\",\"Email\":\"jane@old.com\","
                    + "\"Bio\":null,\"Score\":\"91.2\"}}}"
        };
        try (JsonToolsHandle tools = JsonTools.builder()
                .flatten()
                .separator("::")
                .lowercaseKeys(true)
                .keyReplacement("r'^api_response::user_data::'", "")
                .keyReplacement("first_name", "name")
                .valueReplacement("@old.com", "@new.com")
                .removeEmptyStrings(true)
                .removeNulls(true)
                .removeEmptyObjects(true)
                .removeEmptyArrays(true)
                .autoConvertTypes(true)
                .parallelThreshold(50)
                .numThreads(2)
                .nestedParallelThreshold(200)
                .build()) {
            String[] results = tools.executeBatch(apiBatch);
            System.out.println("   Features: separator, lowercase, 2x keyReplacement, valueReplacement,");
            System.out.println("             4x filtering, autoConvertTypes, parallel tuning, batch");
            for (int i = 0; i < results.length; i++) {
                System.out.println("   [" + i + "]: " + results[i]);
            }
        }
    }
}

File: python/examples/feature_combinations.py
#!/usr/bin/env python3
"""
JSON Tools RS - Feature Combinations (Python)

Curated multi-feature pipelines, building on feature_by_feature.py. Not an
exhaustive combinatorial sweep (the builder has ~10 independent toggles, so a
literal power-set would be 1000+ cases) -- these are realistic groupings of
features commonly used together, plus one "kitchen sink" example exercising
nearly everything at once. Mirrors examples/feature_combinations.rs and the
Java equivalent under jvm/examples/.
"""

import json_tools_rs


def main() -> None:
    print("JSON Tools RS - Feature Combinations")
    print("=====================================\n")

    # 1. separator + lowercase_keys + key_replacement + handle_key_collision
    print("1. separator + lowercase_keys + key_replacement + handle_key_collision")
    data = {"User": {"Full_Name": "John"}, "Admin": {"Full_Name": "Jane"}}
    out = (
        json_tools_rs.JSONTools()
        .flatten()
        .separator("_")
        .lowercase_keys(True)
        .key_replacement("r'^(user|admin)_'", "")
        .handle_key_collision(True)
        .execute(data)
    )
    print(f"   In:  {data}\n   Out: {out}\n")

    # 2. key_replacement + value_replacement together
    print("2. key_replacement + value_replacement")
    data = {"usr_nm": "John", "usr_eml": "john@old.com"}
    out = (
        json_tools_rs.JSONTools()
        .flatten()
        .key_replacement("usr_", "user_")
        .value_replacement("@old.com", "@new.com")
        .execute(data)
    )
    print(f"   In:  {data}\n   Out: {out}\n")

    # 3. All four empty-value filters together
    print(
        "3. remove_empty_strings + remove_nulls + remove_empty_objects + remove_empty_arrays"
    )
    data = {"name": "John", "bio": "", "age": None, "tags": [], "meta": {}}
    out = (
        json_tools_rs.JSONTools()
        .flatten()
        .remove_empty_strings(True)
        .remove_nulls(True)
        .remove_empty_objects(True)
        .remove_empty_arrays(True)
        .execute(data)
    )
    print(f"   In:  {data}\n   Out: {out}\n")

    # 4. Real-world normalization pipeline: replacement + filtering + collision
    print("4. lowercase_keys + key_replacement + filtering + handle_key_collision")
    data = {
        "User_Name": "John",
        "User_Bio": "",
        "Admin_Name": "Jane",
        "Admin_Bio": None,
    }
    out = (
        json_tools_rs.JSONTools()
        .flatten()
        .lowercase_keys(True)
        .key_replacement("r'^(user|admin)_'", "")
        .remove_empty_strings(True)
        .remove_nulls(True)
        .handle_key_collision(True)
        .execute(data)
    )
    print(f"   In:  {data}\n   Out: {out}\n")

    # 5. unflatten + key_replacement + value_replacement
    print("5. unflatten + separator + key_replacement + value_replacement")
    data = {"PREFIX_user_name": "john@OLD.com", "PREFIX_user_age": 30}
    out = (
        json_tools_rs.JSONTools()
        .unflatten()
        .separator("_")
        .key_replacement("PREFIX_user_", "profile_")
        .value_replacement("@OLD.com", "@new.com")
        .execute(data)
    )
    print(f"   In:  {data}\n   Out: {out}\n")

    # 6. normal mode + auto_convert_types + value_replacement + filtering
    print("6. normal + auto_convert_types + value_replacement + remove_empty_strings")
    data = {"user": {"status": "ACTIVE", "note": "", "score": "95.5"}}
    out = (
        json_tools_rs.JSONTools()
        .normal()
        .auto_convert_types(True)
        .value_replacement("r'^ACTIVE$'", "enabled")
        .remove_empty_strings(True)
        .execute(data)
    )
    print(f"   In:  {data}\n   Out: {out}")
    print(
        "   Note: nesting preserved, string replaced, empty note dropped, score converted\n"
    )

    # 7. Batch processing + parallel tuning + type conversion together
    print(
        "7. batch execute + parallel_threshold + num_threads + "
        "nested_parallel_threshold + auto_convert_types"
    )
    batch = [{"id": str(i), "active": "true"} for i in range(150)]
    tools = (
        json_tools_rs.JSONTools()
        .flatten()
        .parallel_threshold(50)
        .num_threads(4)
        .nested_parallel_threshold(100)
        .auto_convert_types(True)
    )
    results = tools.execute(batch)
    print(f"   Processed {len(results)} documents")
    print(f"   Sample: {results[0]}\n")

    # 8. Kitchen sink: (almost) every feature at once, on a realistic messy batch
    print("8. Kitchen sink - every applicable feature combined")
    api_batch = [
        {
            "API_Response": {
                "User_Data": {
                    "First_Name": "John",
                    "Email": "john@old.com",
                    "Bio": "",
                    "Score": "88.5",
                }
            }
        },
        {
            "API_Response": {
                "User_Data": {
                    "First_Name": "Jane",
                    "Email": "jane@old.com",
                    "Bio": None,
                    "Score": "91.2",
                }
            }
        },
    ]
    tools = (
        json_tools_rs.JSONTools()
        .flatten()
        .separator("::")
        .lowercase_keys(True)
        .key_replacement("r'^api_response::user_data::'", "")
        .key_replacement("first_name", "name")
        .value_replacement("@old.com", "@new.com")
        .remove_empty_strings(True)
        .remove_nulls(True)
        .remove_empty_objects(True)
        .remove_empty_arrays(True)
        .auto_convert_types(True)
        .parallel_threshold(50)
        .num_threads(2)
        .nested_parallel_threshold(200)
    )
    results = tools.execute(api_batch)
    print("   Features: separator, lowercase, 2x key_replacement, value_replacement,")
    print("             4x filtering, auto_convert_types, parallel tuning, batch")
    for i, r in enumerate(results):
        print(f"   [{i}]: {r}")


if __name__ == "__main__":
    main()
